orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -165,6 +178,206 @@ def save_as_onnx(self, path):
+                if is_list_input:
+                    input, label = inputs[:-1], inputs[-1]
+                    preds = self.model_(*input)
+                    return self.loss_fn_(preds, label), preds
+                else:
+                    sig = signature(self.model_.forward)
+                    ordered_list_keys = list(sig.parameters.keys())
+
+                    input_dict = {}
+                    for key in sig.parameters.keys():
+                        if key in self.input_names_:
+                            input_dict[key] = inputs[self.input_names_.index(key)]
+
+                    model_out = self.model_(**input_dict)
+                    if self.loss_fn_ is None:
+                        return model_out
+
+                    label = inputs[-1]
+                    preds = model_out
+                    return self.loss_fn_(preds, label), preds
+
+        model = CombineTorchModelLossFn(model, loss_fn, input_names)
+
+        return model
+    
+    def convert_torch_model_loss_fn_to_onnx(self, device, inputs, opset_version=DEFAULT_OPSET_VERSION):
+        input_names = [input_elem[0] for input_elem in self.model_desc.inputs]
+        output_names = [output_elem[0] for output_elem in self.model_desc.outputs]
+        #print(inputs)#print(input_names, output_names)
+        if isinstance(inputs, torch.Tensor):
+            inputs = [inputs]
+        if isinstance(inputs, dict):
+            sample_inputs = [inputs[k.name_].to(device=device) for k in self.model_desc.inputs]
+        elif isinstance(inputs, (list, tuple)):
+            sample_inputs = [input.to(device=device) for i, input in enumerate(inputs) if i < len(self.model_desc.inputs)]
+        else:
+            raise RuntimeError("Unexpected input type. Only torch.Tensor, or dict/list/tuple of torch.Tensor is supported.")
+
+        # pytorch onnx exporter/trace does not try to match argument names.
+        # e.g. for models with optional inputs, it requires all inputs be present.
+        # this is a problem because the model graph depends on inputs provided.
+        model = self._combine_torch_model_with_loss(self._torch_model, self._loss_fn, input_names)
+        
+        model.eval()
+        with torch.no_grad():
+            sample_outputs = model(*sample_inputs)
+        if isinstance(sample_outputs, torch.Tensor):
+            sample_outputs = [sample_outputs]
+        # apply dtypes for input and output
+        for i, sample_input in enumerate(sample_inputs):
+            if i < len(self.model_desc.inputs):
+                self.model_desc.add_type_to_input_description(i, sample_input.dtype)
+        for i, sample_output in enumerate(sample_outputs):
+            if i < len(self.model_desc.outputs):
+                self.model_desc.add_type_to_output_description(i, sample_output.dtype)
+        model.train()
+
+        f = io.BytesIO()
+
+        # Other export options to use(this is for backward compatibility).
```

Comment:
We are not supporting pytorch 1.4, so you can remove from lines 275-284 and just add `training=torch.onnx.TrainingMode.TRAINING` into `torch.onnn._export()` call